File: scala/0206-reverse-linked-list.scala
object Solution {
    def reverseList(head: ListNode): ListNode = {
        if (head == null || head.next == null) {
            return head
        }
        
        val reversedListHead = reverseList(head.next)
        head.next.next = head
        head.next = null
        
        return reversedListHead
    }
}

File: swift/0206-reverse-linked-list.swift
class Solution {
    func reverseList(_ head: ListNode?) -> ListNode? {
        if (head === nil || head?.next === nil) {
            return head
        }
        
        var prev: ListNode? = nil
        var curr = head, next = curr?.next
        
        while curr != nil {
            next = curr?.next
            curr?.next = prev
            prev = curr
            curr = next
        }
        return prev
    }
}